### ISSUE TITLE
BufferUnderflowException in SCQIndexing.moveToEnd()

### ISSUE THREAD

**Issue by mcs6502** [OpenHFT/Chronicle-Queue#559]
I'm getting a runtime exception in SCQIndexing.moveToEnd(). The exception occurs at two different locations in this method: either a BufferUnderflowException in readSkip(), or a DecoratedBufferUnderflowException in readPosition(). It appears that there is a race between updating the writePosition.getVolatileValue() in SCQIndexing and updating the (nonvolatile) writePosition in AbstractBytes base class of MappedBytes.

**Comment by mcs6502** [comment]
```
java.nio.BufferUnderflowException: null
	at net.openhft.chronicle.bytes.MappedBytes.readSkip(MappedBytes.java:394) ~[chronicle-bytes-2.17.8.jar:?]
	at net.openhft.chronicle.bytes.MappedBytes.readSkip(MappedBytes.java:44) ~[chronicle-bytes-2.17.8.jar:?]
	at net.openhft.chronicle.queue.impl.single.SCQIndexing.moveToEnd(SCQIndexing.java:701) ~[chronicle-queue-5.17.3.jar:?]
	at net.openhft.chronicle.queue.impl.single.SingleChronicleQueueStore.moveToEndForRead(SingleChronicleQueueStore.java:205) ~[chronicle-queue-5.17.3.jar:?]
	at net.openhft.chronicle.queue.impl.single.SingleChronicleQueueExcerpts$StoreTailer.optimizedToEnd(SingleChronicleQueueExcerpts.java:1579) ~[chronicle-queue-5.17.3.jar:?]
	at net.openhft.chronicle.queue.impl.single.SingleChronicleQueueExcerpts$StoreTailer.toEnd(SingleChronicleQueueExcerpts.java:1549) ~[chronicle-queue-5.17.3.jar:?]
```


```
net.openhft.chronicle.bytes.util.DecoratedBufferUnderflowException: readPosition failed. Position: 131448 > readLimit: 131384
	at net.openhft.chronicle.bytes.AbstractBytes.readPosition(AbstractBytes.java:190) ~[chronicle-bytes-2.17.8.jar:?]
	at net.openhft.chronicle.bytes.MappedBytes.readPosition(MappedBytes.java:289) ~[chronicle-bytes-2.17.8.jar:?]
	at net.openhft.chronicle.bytes.MappedBytes.readPosition(MappedBytes.java:44) ~[chronicle-bytes-2.17.8.jar:?]
	at net.openhft.chronicle.queue.impl.single.SCQIndexing.moveToEnd(SCQIndexing.java:692) ~[chronicle-queue-5.17.3.jar:?]
	at net.openhft.chronicle.queue.impl.single.SingleChronicleQueueStore.moveToEndForRead(SingleChronicleQueueStore.java:205) ~[chronicle-queue-5.17.3.jar:?]
	at net.openhft.chronicle.queue.impl.single.SingleChronicleQueueExcerpts$StoreTailer.optimizedToEnd(SingleChronicleQueueExcerpts.java:1579) ~[chronicle-queue-5.17.3.jar:?]
	at net.openhft.chronicle.queue.impl.single.SingleChronicleQueueExcerpts$StoreTailer.toEnd(SingleChronicleQueueExcerpts.java:1549) ~[chronicle-queue-5.17.3.jar:?]
```

In the latter case, the delta of 64 (131448 - 131384) appears to match the "len" further down in `int len = Wires.lengthOf(header) + 4`.

**Comment by RobAustin** [comment]
Please provide a failing test, ideally via a pull request.

**Comment by dpisklov** [comment]
@mcs6502 there's no race, the problem is that before toEnd you call toIndex with wrong index, and there was a code which basically says "if I couldn't find the index - I won't allow any more reads from this wire". I fixed that now, and your test passes (I added it to our test suite).

**Comment by hft-team-city** [comment]
Released in [Chronicle-Queue-5.17.5](https://github.com/OpenHFT/Chronicle-Queue/releases/chronicle-queue-5.17.5), [BOM-2.17.196](https://github.com/OpenHFT/OpenHFT/releases/chronicle-bom-2.17.196)

**Comment by hft-team-city** [comment]
Released in [Chronicle-Queue-5.17.5](https://github.com/OpenHFT/Chronicle-Queue/releases/chronicle-queue-5.17.5), [BOM-2.17.225](https://github.com/OpenHFT/OpenHFT/releases/chronicle-bom-2.17.225)

**Comment by hft-team-city** [comment]
Released in [Chronicle-Queue-5.17.5](https://github.com/OpenHFT/Chronicle-Queue/releases/chronicle-queue-5.17.5), [BOM-2.17.241](https://github.com/OpenHFT/OpenHFT/releases/chronicle-bom-2.17.241)
